fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java

```diff
@@ -73,10 +74,21 @@ protected IModLocator.ModFileOrException createMod(Path path, boolean ignoreUnkn
         } else
             return new IModLocator.ModFileOrException(null, new ModFileLoadingException("Invalid mod file found " + path));
 
-        mjm.setModFile(mod);
         return new IModLocator.ModFileOrException(mod, null);
     }
 
+    private static JarMetadata loadMetaFromJar(SecureJar jar, ModJarMetadata mjm) {
+        var automatic = JarMetadata.from(jar, jar.getPrimaryPath());
+
+        if (!jar.moduleDataProvider().findFile(MODS_TOML).isPresent())
+            return automatic;
+
+        if (jar.moduleDataProvider().findFile(MODULE_INFO).isPresent())
+            return automatic;
+
+        return mjm;
```

Comment:
Better not to calculate a module name through SecureModules when we won't use it. The variable name "automatic" is misleading as it could also be from an explicit module-info.java file
```suggestion
        if (jar.moduleDataProvider().findFile(MODULE_INFO).isPresent() || jar.moduleDataProvider().findFile(MODS_TOML).isEmpty())
            return JarMetadata.from(jar, jar.getPrimaryPath());

        return mjm;
```
Could also temporarily cache the result of the mods.toml findFile call inside mjm to avoid checking for mods.toml twice

Reply:
Ah I was gunna make mjm take in the automatic metadata and derrive itself from that because it has the filename sanitization, but decided against it when I limited mjm to mods.toml only so it uses the modid.